gh-117376: Make `Py_DECREF` a macro in ceval.c in free-threaded build

colesbury opened the pull request:
`Py_DECREF` and `PyStackRef_CLOSE` are now implemented as macros in the free-threaded build in ceval.c. There are two motivations;

 * MSVC has problems inlining functions in ceval.c in the PGO build.

 * We will want to mark escaping calls in order to spill the stack pointer in ceval.c and we will want to do this around `_Py_Dealloc` (or `_Py_MergeZeroLocalRefcount` or `_Py_DecRefShared`), not around the entire `Py_DECREF` or `PyStackRef_CLOSE` call.


<!-- gh-issue-number: gh-117376 -->
* Issue: gh-117376
<!-- /gh-issue-number -->

colesbury committed:
gh-117376: Make `Py_DECREF` a macro in ceval.c in free-threaded build

`Py_DECREF` and `PyStackRef_CLOSE` are now implemented as macros in the
free-threaded build in ceval.c. There are two motivations;

 * MSVC has problems inlining functions in ceval.c in the PGO build.

 * We will want to mark escaping calls in order to spill the stack
   pointer in ceval.c and we will want to do this around `_Py_Dealloc`
   (or `_Py_MergeZeroLocalRefcount` or `_Py_DecRefShared`), not around
   the entire `Py_DECREF` or `PyStackRef_CLOSE` call.

DinoV reviewed (approved):
LGTM!

picnixz commented on Include/internal/pycore_stackref.h:
```diff
@@ -191,18 +191,15 @@ PyStackRef_FromPyObjectImmortal(PyObject *obj)
+            if (!PyStackRef_IsDeferred(_close_tmp)) { \
+                Py_DECREF(PyStackRef_AsPyObjectBorrow(_close_tmp)); \
+            } \
+        } while (0)
```
PEP 7 requires to:

- use uppercase letters for macro parameters,
- align line continuations `\`

picnixz reviewed (commented):
Just some comment on using PEP 7, although it's not mandatory for old code to follow it but since you're anyway editing things... Line continuations being aligned could be ignored, but having macro parameters names in upper case is usually helpful when you read the macro definition.

colesbury committed:
Clean-up macro definition.

The `DEALLOC` argument isn't currently used. _Py_DECREF_SPECIALIZED is
the same as `Py_DECREF` in the free-threaded build. This might change in
the future, but keep things simpler for now.

colesbury committed:
Merge branch 'main' into gh-117376-pydecref-macro

colesbury committed:
Format PyStackRef_CLOSE
